[로또 - 웹, DB] 코나스 미션 제출합니다.

Conatuseus opened the pull request:
UI를 짜는데 에러가 자꾸 나서 시간을 많이 소비했습니다.
그래서 급하게 하느라 코드가 너무 더럽네요.
웹과 DB를 하면서 '도메인을 더 잘 짰더라면...' 이라는 생각을 많이 했습니다.
시간이 부족해서 코드가 더럽더라도 잘 봐주시고
많은 지적 부탁드립니다.
3단계끝나고 머지되기 전에 주신 피드백도 반영할 예정입니다.
감사합니다.

Conatuseus committed:
docs : README 작성

Squashed commit of the following:

commit 8fa116350ea3a973842a384509ee118b3af2fc38
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 14:47:19 2019 +0900

    README 작성

Conatuseus committed:
feature : 로또 번호 생성하는 generator 구현

LottoNumberGenerator 테스트 코드 및 클래스 구현
README 수정

Squashed commit of the following:

commit 9d54d5b7c25d6f9a0b2e66fbe9e7eb5f843b0511
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 14:58:45 2019 +0900

    LottoNumberGenerator 테스트 코드 작성 및 구현

Conatuseus committed:
feature : 로또 번호 테스트 코드 추가 및 구현

LottoNumber class 구현
lottoNumber가 올바르지 않으면 예외처리

Squashed commit of the following:

commit c24352c7b30845b423d6dc2bee3ad877a1934712
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:06:17 2019 +0900

    Lotto 자동 생성을 위한 코드 주석처리

commit 74d9751ff91b4ebf3532972a1965ad82e07f38a6
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 15:45:55 2019 +0900

    LottoNumber class 구현

Conatuseus committed:
feature : 로또 자동 생성 클래스 구현

로또 수를 random으로 6개 만들던 것을
lotto 수 46개를 저장해서 섞어서 6개를 가져오는 클래스로 변경

Squashed commit of the following:

commit 8be9a06db29acaf92d9af88128a6d9fc6ea0428e
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:31:13 2019 +0900

    LottoNumberGenerator 삭제. LottoGenerator에서 자동로또 생성하도록하기 위해

commit 3c154e43fc26407a6c0960c742cdad355ed750be
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:30:21 2019 +0900

    불필요한 주석 제거

commit 9465e2f709e3c3dfb90371e23fc97e3ef341902f
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:29:21 2019 +0900

    LottoGenerator 구현

Conatuseus committed:
feature : 로또 클래스 구현

Lotto class 구현
LottoTest 구현
로또 자동생성 클래스 수정
해당 테스트코드 수정

Squashed commit of the following:

commit 5af9b6d7aa1b51740a13e6ef57712e22e7182cf1
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 11:44:11 2019 +0900

    LottoAutoGenerator 수정

commit 40e0ed39818ae0b9cb4896ef6459753a588adacc
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 10:50:47 2019 +0900

    Lotto Test code 수정, LottoAutoGenerator 수정

commit efee64cc4503fca1c1eee5a9159023c49fc8fdf1
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 17:43:21 2019 +0900

    주석처리 코드 삭제

commit 78c0fe180914be34608ccbbab6c8ce0824c088ee
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 17:42:39 2019 +0900

    LottoGenerator -> LottoAutoGenerator 클래스명 변경

commit 45dcdb38e67c0ed61b9cb59dfba76e556e120b14
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 17:01:21 2019 +0900

    TestCode 구현

commit a6ddb59f498f7736764bc44705f8905fb26dbcb6
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue May 28 16:54:12 2019 +0900

    Lotto class 구현

Conatuseus committed:
feature : 지난주 당첨 로또 구현

지난주 당첨 로또인 Winning Lotto 구현
Lottos -> UserLottos 로 이름 변경

Squashed commit of the following:

commit 0075d80fe47ca9b00137c341f3b112f1223a96d3
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 14:53:01 2019 +0900

    불필요한 코드 제거

commit 5a89edb4978faf6079458a3345b64c44e1e87428
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 14:52:33 2019 +0900

    불필요한 코드 제거

commit 4e785e0e82115b36e310b125ed876d9b195f01e2
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 14:50:40 2019 +0900

    WinningLotto 구현

commit 36183158c24860e84dd54c3468ec98a657d56824
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 11:52:05 2019 +0900

    로또 번호 1-45으로 수정 (기존 : 1~46)

commit 3f5f3096d101c5fd085860896309185450f7eb91
Author: conatuseus <conatuseus@gmail.com>
Date:   Wed May 29 11:50:50 2019 +0900

    README 수정

Conatuseus committed:
feature : User 구현

User 클래스 구현
구입 금액 객체화
테스트 코드 작성

Squashed commit of the following:

commit 7a0be6ff09a21eb0dd444fe3d4b96c6f58879fb8
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 16:09:01 2019 +0900

    Money 객체화, User 구현 및 테스트코드 작성

Conatuseus committed:
feature : InputView 구현, WinningLottoParser 구현

InputView 구현
WinningLottoParser 구현
Test 코드 추가

Squashed commit of the following:

commit 3c13f96350ddc6a2536127368d58663453142a28
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:22:00 2019 +0900

    WinningLottoParseTest 구현, 인덱스로 LottoNumber가져오는 메소드 구현

commit f347cf59f3e7d27b6e515978686558311cd0bf75
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:21:44 2019 +0900

    WinningLottoParseTest 구현, 인덱스로 LottoNumber가져오는 메소드 구현

commit a96d1c24772c48814f0900da98e9bf50f410e516
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:10:08 2019 +0900

    Money Test, WinningLottoTest 구현

commit 533958090a198729d01f5661b01004f6c12ebb42
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:02:43 2019 +0900

    InputView, WinningLottoPaser 구현

Conatuseus committed:
feature : 입출력 기능 추가

돈 입출력 기능 추가
자동생성된 로또 출력 기능 추가
당첨 로또 입력 기능 추가

Squashed commit of the following:

commit dfced86a5c7f988fbfd70344fef07bbb21d4e835
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 18:03:31 2019 +0900

    돈 입출력, 자동생성된 로또 출력, 당첨로또 입력 기능 추가

commit dd3e8a4f722fcb9d3fc15ef8b142da45753b0b88
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:28:45 2019 +0900

    README 수정

commit e1c47b8bf7359997872fee1b61619a4c28c8ca3e
Author: conatuseus <conatuseus@gmail.com>
Date:   Thu May 30 17:26:29 2019 +0900

    README 수정

Conatuseus committed:
feature : enum rank 구현

Conatuseus committed:
feature : 당첨 통계 기능, 출력 기능 구현

당첨 통계 기능 추가
출력 기능 추가
코드 정리

Squashed commit of the following:

commit fa76576e6f8d3cc720a0c335c2f018ffbd8772f8
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 16:01:50 2019 +0900

    코드 정리

commit b97dbe563b60e7f303e6a5d74b9129b364f9a1dd
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 15:46:19 2019 +0900

    당첨 통계 구하는 기능, 출력 기능 추가

commit 69cf720b4c94c84f75cafe2af1224b55e10f3412
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 14:17:03 2019 +0900

    수익률 구하는 매서드 추가

Conatuseus committed:
refactor : 상수화, 테스트 코드 수정

Squashed commit of the following:

commit 1e07e2cd135c5d11000539b23c868538e1f8a4fd
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 21:47:36 2019 +0900

    상수화 및 테스트코드 수정

Conatuseus committed:
docs : README 작성

Squashed commit of the following:

commit 08e11f163e77da35a92e8c6c859d131a2c0ead64
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 3 21:51:01 2019 +0900

    Todo 작성

Conatuseus committed:
feature : 수동 로또 수 입력 기능 구현

Squashed commit of the following:

commit f28e725cf50f5ccc8aa28d520622419a17909a1f
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue Jun 4 14:40:29 2019 +0900

    수동 구매 로또 수 입력 기능 추가

Conatuseus committed:
feature : 로또 수동 입력 기능 구현

Squashed commit of the following:

commit 705737fba2dcff5c78c4d65b556bee569213b6f0
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue Jun 4 16:21:53 2019 +0900

    코드 정리

commit 091b4ffa51f8d34f7733777b443a5944b9345cae
Author: conatuseus <conatuseus@gmail.com>
Date:   Tue Jun 4 16:06:41 2019 +0900

    로또 번호 수동 입력 기능 구현

Conatuseus committed:
refactor : 피드백 반영

Lotto의 생성자 파라미터 타입 변경
parser가 1급객체를 반환하도록 수정
불필요한 userTest 삭제
LottoParser 추가 구현

Conatuseus committed:
refactor : 피드백 반영

반환값 일급 컬렉션 적용
List 대신 Set으로 수정
User 객체 삭제

Conatuseus committed:
WebUI

Conatuseus committed:
feature : web ui

Conatuseus committed:
feature : UI 구현

Conatuseus committed:
feat : DB 구현

Conatuseus committed:
sql 파일 추가

pobiconan commented on src/lotto_lotto.sql:
```diff
@@ -0,0 +1,52 @@
+  `numbers` varchar(100) NOT NULL,
+  `round_no` int(10) unsigned NOT NULL,
+  PRIMARY KEY (`id`)
+) ENGINE=InnoDB AUTO_INCREMENT=15 DEFAULT CHARSET=utf8;
```
db에서 export한 파일을 관리하는 것이 아니라 테이블을 생성할 때 사용한 스키마만 관리할 것을 추천한다.

pobiconan commented on src/main/java/lotto/WebUILottoData.java:
```diff
@@ -0,0 +1,114 @@
+
+import java.util.List;
+
+public class WebUILottoData {
```
구현을 위해 필요한 모든 데이터를 한 객체에 구현하기보다 각 요청별로 필요한 데이터를 분리해 구현해 본다.

Conatuseus replied:
dto를 만들어서 분리했습니다!

pobiconan commented on src/main/java/lotto/dao/DBConnection.java:
```diff
@@ -0,0 +1,43 @@
+import java.sql.SQLException;
+
+public class DBConnection {
+    public static Connection getConnection() {
```
이 메소드가 너무 많은 일을 하고 있는 것은 아닌가?
한 가지 일만 하도록 분리해본다.

Conatuseus replied:
메서드를 분리해봤습니다.
미흡한 분리였다면 추가 피드백 부탁드릴게요! 😄

pobiconan commented on src/main/java/lotto/dao/LottoDAO.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoDAO {
```
모든 DAO는 상태 값을 가지지 않는다.
이와 같이 상태 값을 가지지 않는 경우 매번 인스턴스를 생성하지 않도록 강제하기 위해 singleton 패턴을 적용할 수 있다.
singleton 패턴을 적용해 인스턴스 하나를 재사용 가능하도록 구현해 본다.

Conatuseus replied:
싱글톤 패턴을 예전에 공부했었는데 활용할 생각을 하지 않고 있었네요.
사용하면 좋은 예를 알게되서 좋았습니다.
감사합니다!

pobiconan commented on src/main/java/lotto/dao/RoundDAO.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class RoundDAO {
```
DAO 구현부를 보면 많은 중복 코드가 있다.
중복을 제거해 나만의 라이브러리를 만들어 DAO 구현을 좀 더 효과적으로 할 수 있는 방법을 찾아본다.

Conatuseus replied:
DAO 구현부에서 중복 코드를 따로 ConnectionManager라는 곳으로 빼봤는데,
이게 맞는걸까 의심이 드네요...

pobiconan commented on src/main/resources/templates/html/ResultOfRound.html:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<html lang="en">
```
lang이 ko여야하지 않을까?

Conatuseus replied:
확인했습니다!

pobiconan commented on src/test/java/lotto/dao/RoundDAOTest.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Test
+    public void aa(){
+
```
아무런 구현 내용도 없으면 테스트 클래스 삭제한다.

Conatuseus replied:
구현한 줄 알았는데 하지 않았었네요...

pobiconan reviewed (changes requested):
웹과 DB 경험이 많지 않은 상태에서 구현하느라 수고했네요. 👍 
단, 모든 데이터를 한 객체에 모아서 구현하고 있는데 그러지 말고 각 요청마다의 데이터를 여러 객체로 분리해 구현할 것을 추천해요.

Conatuseus committed:
refactor: 피드백 반영

Squashed commit of the following:

commit d37dace74ad41e887a6d2045a8af2e03529db09a
Author: conatuseus <conatuseus@gmail.com>
Date:   Mon Jun 24 01:43:11 2019 +0900

    refactor

pobiconan reviewed (approved):
피드백 반영 잘 했네요. 💯 
로또 마무리 했으니 체스 미션에 집중하세요.
